pandas/core/groupby/generic.py

```diff
@@ -1569,6 +1571,158 @@ def func(df):
 
     boxplot = boxplot_frame_groupby
 
+    def value_counts(
+        self,
+        subset: Sequence[Hashable] | None = None,
+        normalize: bool = False,
+        sort: bool = True,
+        ascending: bool = False,
+        dropna: bool = True,
+    ) -> DataFrame | Series:
+        """
+        Return a Series or DataFrame containing counts of unique rows.
+
+        .. versionadded:: 1.4.0
+
+        Parameters
+        ----------
+        subset : list-like, optional
+            Columns to use when counting unique combinations.
+        normalize : bool, default False
+            Return proportions rather than frequencies.
+        sort : bool, default True
+            Sort by frequencies.
+        ascending : bool, default False
+            Sort in ascending order.
+        dropna : bool, default True
+            Don’t include counts of rows that contain NA values.
+
+        Returns
+        -------
+        Series or DataFrame
+            Series if the groupby as_index is True, otherwise DataFrame.
+
+        See Also
+        --------
+        Series.value_counts: Equivalent method on Series.
+        DataFrame.value_counts: Equivalent method on DataFrame.
+        SeriesGroupBy.value_counts: Equivalent method on SeriesGroupBy.
+
+        Notes
+        -----
+        - If the groupby as_index is True then the returned Series will have a
+          MultiIndex with one level per input column.
+        - If the groupby as_index is False then the returned DataFrame will have an
+          additional column with the value_counts.
+
+        By default, rows that contain any NA values are omitted from
+        the result.
+
+        By default, the result will be in descending order so that the
+        first element of each group is the most frequently-occurring row.
+
+        Examples
+        --------
+        >>> df = pd.DataFrame({
+        ...    'gender': ['male', 'male', 'female', 'male', 'female', 'male'],
+        ...    'education': ['low', 'medium', 'high', 'low', 'high', 'low'],
+        ...    'country': ['US', 'FR', 'US', 'FR', 'FR', 'FR']
+        ... })
+
+        >>> df
+            gender 	education 	country
+        0 	male 	low 	    US
+        1 	male 	medium 	    FR
+        2 	female 	high 	    US
+        3 	male 	low 	    FR
+        4 	female 	high 	    FR
+        5 	male 	low 	    FR
+
+        >>> df.groupby("gender").value_counts(normalize=True)
```

Comment:
isn't this result de-facto always the same as 

```
In [22]: df.assign(dummy=1).groupby(list(df.columns)).size().divide(df.groupby('gender').size(), level='gender')
Out[22]:
gender  education  country
female  high       FR         0.50
                   US         0.50
male    low        FR         0.50
                   US         0.25
        medium     FR         0.25
dtype: float64
```

Reply:
@jreback It is, and that is more or less what the algorithm is doing. Except that it has to cope with:
- `dropna`, `sort`, `as_index `in groupby() (nth() has to re-group the same way to cope with dropna)
- `dropna`, `sort` in value_counts() (applying to different subsets)
- Open-ended list of non-column groupers in groupby()

E.g. 
```
df = pd.DataFrame({"A": [1, 2, 1], "B": [1, 2, 3]})
gb = df.groupby([[4, 5, 4], "A", lambda i: 7 if i == 1 else 8], as_index=False)
gb_all = df.groupby([[4, 5, 4], "A", lambda i: 7 if i == 1 else 8, "B"], as_index=False)
print(gb.size())
print(gb_all.size())
print(gb.value_counts(normalize=True))

   level_0  A  level_2  size
0        4  1        8     2
1        5  2        7     1
   level_0  A  level_2  B  size
0        4  1        8  1     1
1        4  1        8  3     1
2        5  2        7  2     1
   level_0  A  level_2  B  proportion
0        4  1        8  1         0.5
1        4  1        8  3         0.5
2        5  2        7  2         1.0
```
The problem is to compute the grouper for gb_all. I think that my solution is fairly clean considering the possible explosion of edge-cases.
I originally coded the `as_index=False` case separately, but it was very complex. I changed to regrouping with `as_index=True` and then converting from Series to DataFrame with reset_index(). This works great, apart from an ongoing discussion about duplicates (#44755), which I would be happy for you to intervene in!